app/containers/Forms/ProductionFields.tsx

```diff
@@ -1,127 +1,224 @@
-import React from 'react';
+import React, {useState, useCallback} from 'react';
 import {FieldProps} from 'react-jsonschema-form';
 import {createFragmentContainer, graphql} from 'react-relay';
-import {Form, Col} from 'react-bootstrap';
 import {ProductionFields_query} from 'ProductionFields_query.graphql';
+import {JSONSchema6} from 'json-schema';
+import {FormJson} from 'next-env';
 
 interface Props extends FieldProps {
   query: ProductionFields_query;
 }
 
-const ProductionFields: React.FunctionComponent<Props> = ({
+export const ProductionFieldsComponent: React.FunctionComponent<Props> = ({
   formData,
   query,
-  onChange
+  onChange,
+  registry,
+  autofocus,
+  idSchema,
+  errorSchema,
+  formContext,
+  disabled,
+  readonly,
+  schema,
+  uiSchema
 }) => {
-  const changeField = (event, key) => {
+  const {
+    FieldTemplate
+  }: {
+    FieldTemplate: React.FunctionComponent<any>;
+  } = registry as any;
+  // Not using the types definded in @types/react-jsonschema-form as they are out of date
+
+  const handleProductChange = (product: string) => {
+    const productUnits = query.allProducts.edges.find(
+      ({node}) => node.name === product
+    )?.node.units;
+    onChange({
+      ...formData,
+      product,
+      productUnits,
+      additionalData: {productUnits}
+    });
+  };
+
+  const handleAdditionalFieldsChange = (additionalData: any) => {
+    console.log({
+      ...formData,
+      additionalData
+    });
+    // TODO: execute formula here.
     onChange({
       ...formData,
-      [key]: (event.nativeEvent.target as HTMLInputElement).value
+      additionalData
     });
   };
 
+  const getAdditionalDataSchema = useCallback(() => {
+    return query.allProducts.edges.find(
+      ({node}) => node.name === formData.product
+    )?.node?.productFormByProductFormId?.productFormSchema as FormJson;
+  }, [formData.product, query.allProducts.edges]);
+
+  const [productFieldSchema] = useState<JSONSchema6>({
+    ...(schema.properties.product as JSONSchema6),
+    enum: query.allProducts.edges.map(({node}) => node.name)
+  });
+
   return (
     <>
-      <Col xs={12} md={6}>
-        <Form.Group controlId="id.product">
-          <Form.Label>Product</Form.Label>
-          <Form.Control
-            as="select"
-            value={formData.product}
-            onChange={e => {
-              onChange({
-                ...formData,
-                product: (e.nativeEvent.target as HTMLInputElement).value,
-                productUnits: undefined
-              });
-            }}
+      <FieldTemplate
```

Comment:
@hamzajaved By using the registry's `FieldTemplate` and `StringField`, this allows for this component to be reused in both the `Form` and the `ApplicationDetails` pages. It will either render an input or the string value based on what is defined in the `JsonSchemaForm`'s props. Although passing all of these props is a bit verbose, this makes the component fully generic. We can do the same type of change to the Emissions component later, to avoid the current code duplication

Reply:
^ this. 👍 
Can we get this committed as a `jsdoc` or `tsdoc` description block on the class?